include/cooperative_perception/clustering.hpp

```diff
@@ -0,0 +1,230 @@
+    }
+
+    return state_sum_ / std::size(detections_);
+  }
+
+  Point state_sum_;
+  std::unordered_map<Uuid, Detection> detections_;
+};
+
+namespace detail
+{
+struct squared_euclidean_distance_fn
+{
+  template <typename Detection>
+  auto operator()(const Point & point, const Detection & detection) const -> double
+  {
+    double sum{0};
+
+    sum += std::pow(remove_units(point.position_x - detection.state.position_x), 2);
+    sum += std::pow(remove_units(point.position_y - detection.state.position_y), 2);
+    sum += std::pow(remove_units(point.velocity - detection.state.velocity), 2);
+    sum += std::pow(remove_units(point.yaw.get_angle() - detection.state.yaw.get_angle()), 2);
+    sum += std::pow(remove_units(point.yaw_rate - detection.state.yaw_rate), 2);
+
+    return sum;
+  }
+
+  template <typename... Alternatives>
+  auto operator()(Point point, const std::variant<Alternatives...> & detection) const -> double
+  {
+    const Visitor visitor{
+      [this](const Point & p, const CtrvDetection & d) { return this->operator()(p, d); },
+      [this](const Point & p, const CtraDetection & d) { return this->operator()(p, d); },
+      [this](const auto &, const auto &) { std::numeric_limits<double>::max(); }};
+
+    return std::visit(visitor, std::variant<Point>{point}, detection);
+  }
+};
+
+inline constexpr squared_euclidean_distance_fn squared_euclidean_distance{};
+
+template <typename Detection>
+auto make_point(const Detection & detection) -> Point
+{
+  return Point{
+    detection.state.position_x, detection.state.position_y, detection.state.velocity,
+    detection.state.yaw, detection.state.yaw_rate};
+}
+
+template <typename... Alternatives>
+auto make_point(const std::variant<Alternatives...> & detection) -> Point
+{
+  const Visitor visitor{
+    [](const CtrvDetection & detection) {
+      return Point{
+        detection.state.position_x, detection.state.position_y, detection.state.velocity,
+        detection.state.yaw, detection.state.yaw_rate};
+    },
+    [](const CtraDetection & detection) {
+      return Point{
+        detection.state.position_x, detection.state.position_y, detection.state.velocity,
+        detection.state.yaw, detection.state.yaw_rate};
+    },
+    [](const auto &) { throw std::runtime_error("cannot make point from detection"); }};
+
+  return std::visit(visitor, detection);
+}
+
+}  // namespace detail
+
+template <typename Detection>
+[[nodiscard]] auto cluster_detections(std::vector<Detection> detections, double distance_threshold)
+  -> std::vector<Cluster<Detection>>
+{
+  std::vector<Cluster<Detection>> clusters;
+
+  while (!detections.empty()) {
+    Cluster<Detection> cluster;
+
+    const auto origin_detection{detections.back()};
+    const auto origin_point{detail::make_point(origin_detection)};
+
+    cluster.add_detection(origin_detection);
+    detections.pop_back();
+
+    for (auto it{std::begin(detections)}; it != std::end(detections);) {
```

Comment:
Consider changing "it" to a name more specific to detections/objects/etc

Reply:
Using `it` for an iterator in this context is idiomatic for C++ from what I can tell. Do you have suggestions for alternative names?